### PR TITLE
#166240969  Bookmark articles for reading later

### PR DESCRIPTION
**Pull request by EmyRukundo**
**What does this PR do?**

Bookmarking articles for reading later

**Description of Task to be completed?**

Bookmark articles for reading later

**How should this be manually tested?**

yarn test

**Any background context you want to provide?**

**What are the relevant pivotal tracker stories?**

166240969

**Screenshots (if appropriate)**
ezgif com-video-to-gif (1)

### CODE REVIEW

**Review by deschantkn — APPROVED** [review]

**Review by rukundoeric — APPROVED** [review]

File: __tests__/redux/actions/bookmark.test.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import { getBoomarks, bookmark } from '../../../src/redux/actions/article.actions';
+import { bookmarkk } from '../../../__mocks__/data';
+import getArticleReducer from '../../../src/redux/reducers/articles.reducer';
+import { GET_BOOKMARK, BOOKMARK } from '../../../src/redux/actions/types/article.type';
+
+const middleware = [thunk];
+const mockStore = configureStore(middleware);
+const store = mockStore({});
+
+describe('SHOULD TEST BOOKMARKS', () => {
```
Comment (bdushimi) [inline __tests__/redux/actions/bookmark.test.js]: @EmyRukundo the describe is not supposed to have that "SHOULD"

File: __tests__/redux/actions/bookmark.test.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        response: {
+          data: { bookmark: {}, count: 1 },
+        },
+      });
+    });
+    return store.dispatch(bookmark(1)).then(() => {
+      expect(store.getActions().length).toEqual(1);
+    });
+  });
+  test('should not create bookmark', async () => {
```
Comment (bdushimi) [inline __tests__/redux/actions/bookmark.test.js]: @EmyRukundo when won't it be possible to bookmark an article? this test is not descriptive enough.

File: __tests__/redux/actions/bookmark.test.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+          data: { bookmark: 'error' },
+        },
+      });
+    });
+    return store.dispatch(bookmark(1)).then(() => {
+      expect(store.getActions().length).toEqual(1);
+    });
+  });
+});
+describe('Reducer test bookmarking', () => {
```
Comment (bdushimi) [inline __tests__/redux/actions/bookmark.test.js]: @EmyRukundo wouldn't be better to separate actionsTests & reducerTests into different testsFolder/testFiles?

File: __tests__/redux/actions/bookmark.test.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  const initialState5 = {
+    Bookmarks: [],
+    bookmark: {},
+  };
+  it('should return an object', () => {
+    const getbookmarks = getArticleReducer(initialState5, {
+      type: GET_BOOKMARK,
+      payload: bookmarkk[0],
+    });
+    expect(typeof getbookmarks).toBe('object');
```
Comment (bdushimi) [inline __tests__/redux/actions/bookmark.test.js]: @EmyRukundo you're only testing the data types, you should test also the actual data.

File: __tests__/redux/actions/bookmark.test.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  });
+  it('should return an object', () => {
+    const bookmarkArticle = getArticleReducer(initialState5, {
+      type: BOOKMARK,
+      payload: bookmarkk[0],
+    });
+    expect(typeof bookmarkArticle).toBe('object');
```
Comment (bdushimi) [inline __tests__/redux/actions/bookmark.test.js]: @EmyRukundo you're only testing the data types, you should test also the actual data.

File: __tests__/redux/actions/bookmark.test.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      const request = moxios.requests.mostRecent();
+      request.respondWith({
+        status: 404,
+        response: {
+          data: { bookmark: 'error' },
+        },
+      });
+    });
+    return store.dispatch(bookmark(1)).then(() => {
+      expect(store.getActions().length).toEqual(1);
```
Comment (bdushimi) [inline __tests__/redux/actions/bookmark.test.js:60]: @EmyRukundo you're not really testing if the expected action (bookmark) has been dispatched successfully! It should be something like this:  expect(store.getActions()).toEqual(expectedAction);

File: src/components/pages/ReadArticle.js [hunk 7]
```diff
@@ -14,27 +23,79 @@ class ReadArticle extends Component {
...
+    isBookmarked: false,
+    isProfileRequested: false,
+    isPreviousBookmarked: false,
+    redirect: false,
   };
 
   componentWillMount() {
     const { slug } = this.props.match.params;
     this.props.getArticle(slug);
+    this.props.getBoomarks(slug);
```
Comment (bdushimi) [inline src/components/pages/ReadArticle.js]: @EmyRukundo why does getBookmarks function need a slug as parameter? Additionally, the function signature in src/redux/actions/article.actions.js does not require the slug parameter.

File: src/components/pages/ReadArticle.js [hunk 8]
```diff
@@ -14,27 +23,79 @@ class ReadArticle extends Component {
...
+      if (
+        this.isThisSlugBookmarked(slug, newProps.myBookmarks)
+        && !this.state.isPreviousBookmarked
+      ) {
+        this.setState({ isBookmarked: true, isPreviousBookmarked: true });
+      }
+    }
   }
 
+  handleClick = () => {
```
Comment (bdushimi) [inline src/components/pages/ReadArticle.js]: @EmyRukundo make function names more descriptive and readable enough. handleClick, yes, but which click and on which component/button?

File: src/redux/reducers/articles.reducer.js [hunk 9]
```diff
@@ -15,7 +18,7 @@ export default (state = initialState, action) => {
         ...state,
         articles: payload,
       };
-    case GET_ARTICLE:
+    case GET_ONE_ARTICLE:
```
Comment (bdushimi) [inline src/redux/reducers/articles.reducer.js]: @EmyRukundo this is the same as GET_ARTICLE right?

**Review by bdushimi — CHANGES_REQUESTED** [review]

File: src/components/pages/ReadArticle.js (hunk 8 above)
Comment (EmyRukundo) [reply src/components/pages/ReadArticle.js]: Thanks for the feedback, I am going to correct it

File: src/components/pages/ReadArticle.js (hunk 7 above)
Comment (EmyRukundo) [reply src/components/pages/ReadArticle.js]: thanks, corrected it

File: __tests__/redux/actions/bookmark.test.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+});
+describe('Reducer test bookmarking', () => {
+  const initialState5 = {
+    Bookmarks: [],
+    bookmark: {},
+  };
+  it('should return an object', () => {
+    const getbookmarks = getArticleReducer(initialState5, {
+      type: GET_BOOKMARK,
+      payload: bookmarkk[0],
```
Comment (bdushimi) [inline __tests__/redux/actions/bookmark.test.js]: @EmyRukundo little typo bookmark

File: __tests__/redux/actions/socialLogin.test.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    };
+    moxios.wait(() => {
+      const request = moxios.requests.mostRecent();
+      request.respondWith({
+        status: 200,
+        response: expectedResults,
+      });
+    });
+    return Store.dispatch(socialLogin(accessToken, provider)).then(() => {
+      expect(Store.getActions()).toBeDefined();
```
Comment (bdushimi) [inline __tests__/redux/actions/socialLogin.test.js]: @EmyRukundo what exactly are you testing here?

**Review by bdushimi — CHANGES_REQUESTED** [review]

File: __tests__/redux/actions/bookmark.test.js (hunk 1 above)
Comment (EmyRukundo) [reply __tests__/redux/actions/bookmark.test.js]: Thanks! I will correct it

File: __tests__/redux/actions/bookmark.test.js (hunk 2 above)
Comment (EmyRukundo) [reply __tests__/redux/actions/bookmark.test.js]: Thanks! I am going to work on it

File: __tests__/redux/actions/bookmark.test.js (hunk 3 above)
Comment (EmyRukundo) [reply __tests__/redux/actions/bookmark.test.js]: I am going to work on it

File: __tests__/redux/actions/bookmark.test.js (hunk 4 above)
Comment (EmyRukundo) [reply __tests__/redux/actions/bookmark.test.js]: Thanks, I will work on it

File: __tests__/redux/actions/bookmark.test.js (hunk 6 above)
Comment (EmyRukundo) [reply __tests__/redux/actions/bookmark.test.js:60]: Thanks, I will work on it

File: src/redux/actions/socialLogin.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+import { BACKEND_URL } from '../../utils/constants';
+
+
+/**
+ * @returns {*} dispatch
+ * @param {string} accessToken
+ * @param {string} provider
+ */
+
+export const socialLogin = (accessToken, provider) => async (dispatch) => {
```
Comment (eliemugenzi) [inline src/redux/actions/socialLogin.js]: Where are you calling these actions?

File: src/redux/actions/socialLogin.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    });
+  }
+};
+
+/**
+ * @returns {*} dispatch
+ * @param {string} token
+ * @param {string} username
+ */
+export const loginWithTwitter = (token, username) => (dispatch) => {
```
Comment (eliemugenzi) [inline src/redux/actions/socialLogin.js]: Are these actions being called anywhere in the project?

File: src/redux/actions/socialLogin.js [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  const user = {
+    token,
+    username,
+  };
+  dispatch({
+    type: SUBMIT_LOGIN,
+    payload: user,
+  });
+};
+export const loginWithGoogle = (token, username) => (dispatch) => {
```
Comment (eliemugenzi) [inline src/redux/actions/socialLogin.js]: The same here

File: src/redux/actions/socialLogin.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+  const user = {
+    token,
+    username,
+  };
+  dispatch({
+    type: SUBMIT_LOGIN,
+    payload: user,
+  });
+};
+export const loginWithFacebook = (token, username) => (dispatch) => {
```
Comment (eliemugenzi) [inline src/redux/actions/socialLogin.js]: The same here

File: src/utils/socialLoginTesla.js [hunk 16]
```diff
@@ -1,12 +1,14 @@
...
+import { BACKEND_URL } from './constants';
+
 
 const { NODE_ENV } = process.env;
-const token = localStorage.getItem('ACCESS_TOKEN') || null;
+const token = sessionStorage.getItem('ACCESS_TOKEN') || null;
 
 export default NODE_ENV === 'test' ? axios : axios.create({
-  baseURL: process.env.REACT_APP_BASE_URL,
+  baseURL: BACKEND_URL,
```
Comment (eliemugenzi) [inline src/utils/socialLoginTesla.js:11]: Where are you using this file?

**Review by eliemugenzi — CHANGES_REQUESTED** [review]

File: __tests__/redux/actions/socialLogin.test.js [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        response: expectedResults,
+      });
+    });
+    Store.dispatch(socialLogin(accessToken, provider));
+    const action = Store.getActions();
+    expect(action[0]).toEqual(expectedResults[0]);
+    return Store.dispatch(socialLogin(accessToken, provider)).then(() => {
+    });
+  });
+  it('should let user login with twitter', () => {
```
Comment (eliemugenzi) [inline __tests__/redux/actions/socialLogin.test.js]: You shouldn't have this file since you are not using any action for social login

**Review by eliemugenzi — CHANGES_REQUESTED** [review]

**Review by eliemugenzi — APPROVED** [review]
Good job @EmyRukundo

**Comment by Gidraf** [comment]
@EmyRukundo Kindly add a screenshot of the functionality on this PR so that I can easily see how the feature works.

**Pull request merged**